qutip/solver/ode/wrapper.pyx

```diff
@@ -0,0 +1,164 @@
+#cython: language_level=3
+
+import numpy as np
+from .._solverqevo cimport SolverQEvo
+from ...core cimport data as _data
+from ...core.data.norm cimport frobenius_dense
+from ...core.data.norm import frobenius
+
+cdef class QtOdeFuncWrapper:
+    def __init__(self, SolverQEvo evo):
+        self.evo = evo
+
+    cpdef void call(self, QtOdeData out, double t, QtOdeData y):
+        if out.inplace:
+            self.evo.mul_dense(t, y.data(), out.data())
+        else:
+            out.set_data(self.evo.mul_data(t, y.data()))
+
+
+cdef class PythonFuncWrapper(QtOdeFuncWrapper):
+    cdef object f
+
+    def __init__(self, object f):
+        self.f = f
+
+    cpdef void call(self, QtOdeData out, double t, QtOdeData y):
+        out.copy(QtOdeData(self.f(t, y.raw())))
+
+
+cdef class QtOdeData:
```

Comment:
I don't understand what these `Data` classes are for and ideally it would be nice to make them go away somehow and use more ordinary things like Qobjs somehow.

Reply:
it looks like they're meant to be dispatched operations over the core `qutip.core.data` types (`CSR` and `Dense`) at the C level.  I assume it's being done here to avoid dynamic allocations in tight loops, but it really just doesn't feel right at all.

`copy` is already implemented on data types, and in addition `raw`, `data` and `set_data` would all be meaningless if the `QtOdeData` type was dropped.  The two numpy-like types can simply be dropped and replaced with conversions up to `Dense` - there shouldn't be any meaningful data kicking around inside QuTiP which _isn't_ in our `Data` formats anyway.

`iadd`, `imul` and `izero` could just be defined on the `Data` types themselves - these data types are meant to help us in QuTiP, not cause us to write huge amounts of boilerplate to wrap them for every specific instance.  It's nice if it's easy to implement a new data-layer type, but that should be a secondary concern.

While we want to avoid Python-space calls while in C mode, I think it's fine to "hard-code" C-level dispatchers that operate _purely_ on single types - the cost of having (defined in `data.norm`,
```cython
cdef double frobenius_(Data data):
    dtype = type(data)
    if dtype == CSR:
        return frobenius_csr(data)
    if dtype == Dense:
        return frobenius_dense(data)
    return np.norm(data.to_array())
```
isn't that high.  It's not ideal, but short of dropping all the way to C++ to get more performant run-time type-dispatch, I'm not sure there's a better way to do it in Cython.

Reply:
I left the support for general python function beside QobjEvo and numpy array beside Data. Do we want to keep that? Or stick with something that work only inside Qutip?  If we drop more general support, we can remove these wrapper.

And yes, I spent a lot of time trying to work around using Data and the Dispatcher. I went with light wrapper around Data since I felt force to use the dispatcher otherwise... Now that "hard-code" C-level dispatchers are fine, I will update with that.